safehttp/mux.go

```diff
@@ -0,0 +1,114 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package safehttp
+
+import (
+	"net/http"
+)
+
+// TODO: add the missing methods
+const (
+	// HTTP GET request
+	MethodGet = "GET"
+	// HTTP Post request
+	MethodPost = "POST"
+)
+
+// ServeMux is a safe HTTP request multiplexer that wraps http.ServeMux.
+// It matches the URL of each incoming request against a list of registered
+// patterns and calls the handler for the pattern that most closely matches the
+// URL.
+//
+// The multiplexer contains a list of allowed domains that will be matched
+// against each incoming request. A different handler can be specified for every
+// HTTP method supported at a registered pattern.
+type ServeMux struct {
+	mux        *http.ServeMux
+	domains    map[string]bool
+	dispatcher Dispatcher
+
+	// Maps user-provided patterns to combined handlers which encapsulate
+	// multiple handlers, each one associated with an HTTP method.
+	handlers map[string]combinedHandler
+}
+
+// NewServeMux allocates and returns a new ServeMux
+// TODO(@mattias, @mara): make domains a variadic of string **literals**.
+func NewServeMux(dispatcher Dispatcher, domains ...string) *ServeMux {
+	d := map[string]bool{}
+	for _, host := range domains {
+		d[host] = true
+	}
+	return &ServeMux{
+		mux:        http.NewServeMux(),
+		domains:    d,
+		dispatcher: dispatcher,
+		handlers:   map[string]combinedHandler{},
+	}
+}
+
+// Handle registers a handler for the given pattern and method. If another
+// handler is already registered for the same pattern and method, Handle panics.
+func (m *ServeMux) Handle(pattern string, method string, h Handler) {
+	ch, ok := m.handlers[pattern]
+	if !ok {
+		ch := combinedHandler{
+			h:       make(map[string]Handler),
```

Comment:
Note, optional change here:

```suggestion
			h:       map[string]Handler{method: h},
```

And remove line 72